ambry-cloud/src/main/java/com.github.ambry.cloud/azure/AzureCloudBlobCryptoService.java

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright 2019 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.cloud.azure;
+
+import com.github.ambry.cloud.CloudBlobCryptoService;
+import com.github.ambry.router.CryptoService;
+import com.github.ambry.router.KeyManagementService;
+import com.github.ambry.utils.Crc32;
+import java.nio.ByteBuffer;
+import java.security.GeneralSecurityException;
+
+
+/**
+ * Implementation of CloudBlobCryptoService, encrypts byte buffers by
+ * 1. generating a random encryption key
+ * 2. encrypting that random key with a key from KeyManagementService
+ *    that is associated with the given context string
+ * 3. encrypting a data byte buffer with the generated random key
+ * 4. returning a byte buffer containing the serialized encrypted key
+ *    and encrypted data (see {@link EncryptedKeyAndEncryptedData} below
+ *    for serialization description
+ */
+public class AzureCloudBlobCryptoService implements CloudBlobCryptoService {
```

Comment:
You can leave Azure out of the name (and package) since it has nothing Azure specific.
I'm concerned that we're adding another message format (EncryptedKeyAndEncryptedData).  How much work would it be to use the same one used by the router?
By not deserializing the blob properties, we are missing some opportunities to avoid encryption on blobs that are already encrypted but the blobId bit was not set.  (I think this only applies to V3 blobs so maybe not a large set.)

Reply:
Yeah I can name it something different.

My concern with using the current router encryption scheme is that to do that, I'd have to
1. Parse the message buffer into its component headers and fields
2. Adjust all the offsets in the header message as the blob encryption field is among the first fields, and adjust any related calculated CRCs
3. Live with weird inconsistent data states like the blob encryption bit in the blob ID being out-of-sync with the actual encryption status beyond Version 3.

The encryption tool would also be limited to use on specific versions of blobs, and would have to change or be tested whenever new versions of blobs (or new versions of components of blobs like message headers) are introduced.  By decoupling cloud backup encryption from whatever the blob content is, it becomes a significantly less complex and more modular, independent, and cleaner component (you could use it to encrypt-decrypt any ByteBuffer, not just Ambry blobs)

I thought about that when writing it (we'd be potentially encrypting V3s that actually were already encrypted but weren't indicated as such according to their blob ID encryption bit) but I felt that 
1. we haven't been making any more new V3s for a while, so this won't be a recurring issue
2. the scale of the affected "double" encrypted blobs would be limited

Talking to other engineers on the team, it sounds like it may not even be a problem as the issue was contained before it was deployed so that's something I'll confirm

Reply:
Looked at the code, looks like it's possible to re-use existing code for un-encrypted blobs and use same format like router.

For deserialization: 
We have `MessageFormatRecord::deserializeBlobAll`, it's pretty neat to use. 

For encryption:
just encrypt it like `encryptJob`.  

For actual write:
After encryption, you can construct `PutMessageFormatInputStream`(to serialize) and write it to CloudStore. (`handlePutRequest` is an example).

Reply:
@zzmao it is indeed possible.  It is also indeed more error prone and more likely to involve unindented consequences by mutating the data.  I am very much aware of `MessageFormatRecord` and `PutMessageFormatInputStream`; I wrote `BlobIdTransformer` which uses these classes to parse and mutate blob records for the rewrite.  It is through that experience that I know that bit fiddling with existing data can be very problematic and risky.

This current encryption scheme is independent of the data that it actually encrypts, and thus is independent of any version changes, future or past, that may happen with any components of `MessageFormatRecord` or `PutMessageFormatInputStream`.  I talked with @cgtz and @pnarayanan too about this and they agreed with me that having an additional header that isolates this encryption process from other data manipulation processes is the way to go.

Reply:
I see. Mutating data is indeed error prone.